app/controllers/doorkeeper/authorizations_controller.rb

```diff
@@ -49,10 +50,21 @@ def render_error
       end
     end
 
+    def can_authorize_response?
+      matching_token? && pre_auth.client.application.confidential? && Doorkeeper.config.custom_access_token_attributes.empty?
+    end
```

Comment:
Could we change the order of this check? 

```suggestion
    def can_authorize_response?
      Doorkeeper.config.custom_access_token_attributes.empty? && pre_auth.client.application.confidential? && matching_token?
    end
```

Like this we don't have to do the whole `matching_token?` processing if one of the two boolean checks fails.